### PR TITLE
Add support for lowering mhlo.fft to Linalg.

### PR DESCRIPTION
**Pull request by hanhanW**
Only RFFT type is supported, which means the inputs are all real
numbers. Since we do not handle complex numbers in IREE and mhlo.fft
returns a complex number, we have to use mhlo.complex op to pack real
part and imaginary part together. And rely on canonicalization patterns
to fold it away. They will be folded away because we ran
`PopulateComplexLoweringPatterns` in HLOToHLOPreprocessing pass, if
there are complex number computation, mhlo.real/imag will be created.

The implementation is not FFT algorithm. It follows DFT definition to
support naive solution. See definition
https://en.wikipedia.org/wiki/Discrete_Fourier_transform

Fixes https://github.com/google/iree/issues/5299

### CODE REVIEW

**Comment by hanhanW** [comment]
I'd like to add end-to-end testing later. It only works on Linalg on tensors, but current iree/test/e2e/*/BUILD is a bit messy.

**Comment by hanhanW** [comment]
> I'd like to add end-to-end testing later. It only works on Linalg on tensors, but current iree/test/e2e/*/BUILD is a bit messy.

I think I was wrong, I probably can add the e2e test.

**Comment by hanhanW** [comment]
This is really weird, I can't fix the tests.

I hacked into printf representation, and got something weird.

https://github.com/google/iree/blob/abab8572e3bf3f81dca6fcd9de18e7660a0c2345/iree/hal/string_util.cc#L394-L396

```
265: Expected equality of these values. Contents does not match.
265:   lhs:
265:     17xf32=666.84601 -0.00001 -590.16931 -0.00000 593.44849 -0.00009 -579.52881 -0.00041 629.95404 0.00061 -567.11255 -0.00021 591
.75140 0.00042 -583.18927 -0.00062 630.84601
265:   rhs:
265:     17xf32=666.84601 -0.00001 -590.16931 -0.00000 593.44849 -0.00009 -579.52881 -0.00041 629.95404 0.00061 -567.11255 -0.00021 591
.75140 0.00042 -583.18927 -0.00062 630.84601
```

They look identical to me. I even just copy-paste the result, and can still get some similar failures.

**Comment by hanhanW** [comment]
oh, I'm so stupid >~<
It should be `expect_almost_eq_const`, not `expect_eq_const`

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp [hunk 1]
```diff
@@ -182,6 +182,120 @@ struct ConcatenateOpConversion
...
+                        bool isRealPart) {
+  int rank = matrixType.getRank();
+  auto initTensor = b.create<linalg::InitTensorOp>(
+      loc, /*dyn_size=*/ValueRange{}, matrixType.getShape(),
+      matrixType.getElementType());
+  SmallVector<StringRef, 2> loops(rank, getParallelIteratorTypeName());
+  // scale = 2 * pi / N
+  double kScale = 2 * acos(-1) / matrixType.getDimSize(0);
+  if (!isRealPart) kScale = -kScale;  // -sin(x) = sin(-x)
+  return b
```
Comment (asaadaldien) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp:203]: You can instead turn this into `const dense<[DFT_VALUES]>` and generate the values here. matrixType is already ranked and all you need to get `DFT_VALUES` is the rank.

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp [hunk 2]
```diff
@@ -203,6 +317,18 @@ struct ConvertHLOToLinalgOnTensorsPass
 
     ConversionTarget target(getContext());
     target.addIllegalDialect<mhlo::MhloDialect>();
+
+    // TODO(hanchung): Do it in a cleaner way.
+    // We don't see complex types in codegen. This assumes that we run
```
Comment (asaadaldien) [inline iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp:322]: One idea is to change the pattern above to look for users that are `mhlo.abs`and replace it a `linalg.generic` that takes the real and imaginary parts and compute the absolute.

Comment (hanhanW) [reply iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp:322]: Good point, I think we are already doing it in this way. There are patterns to rewrite `mhlo.abs` (which takes complex typese) to `mhlo.real + mhlo.imag` to compute the magnitude in HLOToHLOPreprocessing pass.

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp (hunk 1 above)
Comment (hanhanW) [reply iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp]: Good point! I was trying to handle dynamic shapes as well, but it's actually not supported and we're not close to that case.  Let's generate the matrix here, it also addresses accuracy issues I hit.

**Review by asaadaldien — APPROVED** [review]

File: iree/test/e2e/xla_ops/BUILD [hunk 3]
```diff
@@ -140,7 +140,7 @@ iree_check_single_backend_test_suite(
 
 iree_check_single_backend_test_suite(
     name = "check_dylib-llvm-aot_dylib",
-    srcs = DYLIB_AND_VULKAN_CHECK_TESTS,
+    srcs = DYLIB_AND_VULKAN_CHECK_TESTS + ["fft.mlir"],  #TODO(hanchung): clean this up, like VMLA test suite.
```
[the remark this answers is not in the document]
Comment (hanhanW) [reply iree/test/e2e/xla_ops/BUILD:143]: linalg.vecmat lowering is not supported, tracking at https://github.com/google/iree/issues/5444

File: iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp [hunk 4]
```diff
@@ -182,6 +182,115 @@ struct ConcatenateOpConversion
...
+/// Creats coefficients based on DFT definition, see
+/// https://en.wikipedia.org/wiki/Discrete_Fourier_transform
+Value getDFTMatmulCoeff(OpBuilder b, Location loc, RankedTensorType matrixType,
+                        bool isRealPart) {
+  assert(matrixType.getRank() == 2 && "expected 2D matrix");
+
+  auto initTensor = b.create<linalg::InitTensorOp>(loc, matrixType.getShape(),
+                                                   matrixType.getElementType());
+  // scale = 2 * pi / N
+  double scale = 2 * acos(-1) / matrixType.getDimSize(0);
```
[the remark this answers is not in the document]
Comment (hanhanW) [reply iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp]: Done. I did not know this. :D

**Pull request merged**